[JIT] Add a pass for annotating graph with input types derived from sample inputs.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57074 [TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`.
* #57382 [TensorExpr] Remove `dtype_` and add `buf_` fields to `CodeGen::BufferArg`.
* #57169 [JIT] Add a pass for removing a first (self) argument from a graph if it is unused.
* **#57076 [JIT] Add a pass for annotating graph with input types derived from sample inputs.**

This pass is intended to be used in conjunction with shape propagation
pass: first we use sample inputs to specify shape info for graph inputs
and then we run shape-prop to infer shapes of intermediate values in the
graph.

Differential Revision: [D28048290](https://our.internmc.facebook.com/intern/diff/D28048290)

Commit:
[JIT] Add a pass for annotating graph with input types derived from sample inputs.

This pass is intended to be used in conjunction with shape propagation
pass: first we use sample inputs to specify shape info for graph inputs
and then we run shape-prop to infer shapes of intermediate values in the
graph.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add a pass for annotating graph with input types derived from sample inputs."


This pass is intended to be used in conjunction with shape propagation
pass: first we use sample inputs to specify shape info for graph inputs
and then we run shape-prop to infer shapes of intermediate values in the
graph.

[ghstack-poisoned]

ZolotukhinM commented on test/test_tensorexpr_pybind.py:
```diff
@@ -83,14 +83,12 @@ def f(a, b, c):
         """
         graph = torch._C.parse_ir(graph_str)
 
-        with kernel_arena_scope():
```
`TensorExprKernel` sets up its own kernel arena, so we don't need to do this.

Commit:
Update on "[JIT] Add a pass for annotating graph with input types derived from sample inputs."


This pass is intended to be used in conjunction with shape propagation
pass: first we use sample inputs to specify shape info for graph inputs
and then we run shape-prop to infer shapes of intermediate values in the
graph.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add a pass for annotating graph with input types derived from sample inputs."


This pass is intended to be used in conjunction with shape propagation
pass: first we use sample inputs to specify shape info for graph inputs
and then we run shape-prop to infer shapes of intermediate values in the
graph.

Differential Revision: [D28048290](https://our.internmc.facebook.com/intern/diff/D28048290)

[ghstack-poisoned]

eellison reviewed (commented):
Is there a reason to do this on the TensorExpr kernel instead of the JIT Graph? Mobilenet we will probably be able to clean up the exceptions just by propagating shapes without inputs, but in general we might want to use the shape inputs on JIT Graph to clean it up and then lower to TE

ZolotukhinM commented:
> Is there a reason to do this on the TensorExpr kernel instead of the JIT Graph? Mobilenet we will probably be able to clean up the exceptions just by propagating shapes without inputs, but in general we might want to use the shape inputs on JIT Graph to clean it up and then lower to TE

The pass is independent on TE, you're right. I just wasn't sure where to put it and wasn't sure such a small pass justifies adding a new file into `jit/passes`. I could move it there if you prefer. As for the graph cleanup - we still need to concrete shapes to be able to lower to TE, and this pass is intended to be used for that (as a first step to specify start conditions which then will be propagated by the shape inference pass).

eellison commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2873,13 +2901,13 @@ Tensor* TensorExprKernel::convertOutputToCorrectStrides(torch::jit::Value* v) {
 
   auto dims = dimsFromSizes(sizesForValue(v));
   // We need to convert the output tensor so that its values are layed
-  // so that whene viewed from the output strides the values are correct.
```
i gotta start using grammarly or something haha

eellison reviewed (approved):
LGTM we may need to move this to general JIT graph as some point, or this might get replaced by input descriptors. All of these initial APIs are somewhat temporary / internal anyway.

Commit:
Update on "[JIT] Add a pass for annotating graph with input types derived from sample inputs."


This pass is intended to be used in conjunction with shape propagation
pass: first we use sample inputs to specify shape info for graph inputs
and then we run shape-prop to infer shapes of intermediate values in the
graph.

Differential Revision: [D28048290](https://our.internmc.facebook.com/intern/diff/D28048290)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add a pass for annotating graph with input types derived from sample inputs."


This pass is intended to be used in conjunction with shape propagation
pass: first we use sample inputs to specify shape info for graph inputs
and then we run shape-prop to infer shapes of intermediate values in the
graph.

Differential Revision: [D28048290](https://our.internmc.facebook.com/intern/diff/D28048290)

[ghstack-poisoned]
